src/sagemaker/jumpstart/types.py

```diff
@@ -944,6 +1003,273 @@ def to_json(self) -> Dict[str, Any]:
                     json_obj[att] = cur_val
         return json_obj
 
+
+class JumpStartPresetComponent(JumpStartMetadataBaseFields):
+    """Data class of JumpStart config component."""
+
+    slots = ["component_name", "component_override_fields"]
+
+    # List of fields that is not allowed to override to JumpStartMetadataBaseFields
+    # TODO: finalize the fields that can be overrided
+    OVERRIDING_DENY_LIST = []
+
+    __slots__ = slots + JumpStartMetadataBaseFields.__slots__
+
+    def __init__(self, component: Optional[Dict[str, Any]], spec: Optional[Dict[str, Any]]):
+        """Initializes a JumpStartInferenceComponent object from its json representation.
+
+        Args:
+            component (Dict[str, Any]):
+                Dictionary representation of the config component that can override to the metadata base fields.
+            spec (Dict[str, Any]):
+                Dictionary representation of the original metadata base fields.
+        """
+        super().from_json(spec)
+        self.component_name: str = component["component_name"]
+        self.component_overrides: JumpStartMetadataBaseFields = JumpStartMetadataBaseFields(
+            component
+        )
+
+    def to_json(self) -> Dict[str, Any]:
+        """Returns json representation of JumpStartInferenceComponent object."""
+        json_obj = {att: getattr(self, att) for att in self.__slots__ if hasattr(self, att)}
+        return json_obj
+
+    @staticmethod
+    def _override_merge(source: Dict[str, Any], override: Dict[str, Any]):
+        """
+        Deep merge two dictionaries recursively, with override overriding source.
+        """
+        for key, value in source.items():
+            if (
+                key not in JumpStartPresetComponent.OVERRIDING_DENY_LIST
+                and key in override
+                and isinstance(value, dict)
+                and isinstance(override[key], dict)
+            ):
+                JumpStartPresetComponent._override_merge(value, override[key])
```

Comment:
generally not a fan of recursion, you can overflow the stack. what was wrong with the other solution i commented?

Reply:
i guess this is a nit, since our metadata isn't nested so much it'd cause the process to crash